Milestone 1.1 : Update the documentation for End-To-End tests and Debug End-To-End wiki page. (#7)
oppia/oppia-web-developer-docs: Guide-to-migrate-e2e-tests.md

| Navigate to the given destination        |`browser.get(url)`                        |`browser.url(url)`     |
| Get the url of the currently opened website.|`browser.getCurrentUrl()`              |`browser.getUrl()`     |
| Get an attribute from a DOM-element based on the attribute name.|`element.getAttribute(attributeName)`|`$(selector).getAttribute(attributeName)`|
| Wait until a given condition is fulfilled.|`browser.wait(condition, time, optional_msg)`|`browser.waitUntil(condition, { timeout, timeoutMsg, interval })`|
| Send a sequence of keystrokes to the active element|`element.sendKeys(keys)` |`element.setValue(keys)`     |
| Selected DOM-Element is clickable, visible, and exists.|`until.elementToBeClickable(element)` |`$(selector).isClickable()`|
|Get the value of a `<textarea>`, `<select>` or `<input>` found by given selector|`element.getText()` |`$(selector).getValue()`|
|Clear a `<textarea>` or `<input>` element’s value   |`element.clear()` |   `$(selector).clearValue()`  |

For more details please visit [webdriverio's official documentation](https://webdriver.io/docs/gettingstarted).

## Points to note while migrating

> 

Reviewer: I think it would be good to add the Hybrid state file too because the contributors need to be understand the hybrid state too before migrating. WDYT?

Author: Done.

Reviewer: Resolved

1. The basic utility files i.e. forms.js, waitFor.js, action.js, general.js, user.js will already be migrated for the ease of contributors.

   | Files                                      | Expected Migration Date            |
   | -------------------------------------------|  ----------------------------------|
   |  waitFor.js, action.js, general.js, user.js|   2 July 2022                      |
   | forms.js                                   |   20 July 2022                     |

> 2. For more information on how to add new tests or modify the existing tests please refer to [Write E2E tests](WebdriverIO.md#run-e2e-tests) section.

Reviewer: Most probably you are referring to webdriverio. It would be good if you mention that too above.

Author: Done.

Reviewer: Resolved

## Example migration

**Protractor**

  ```js
  var until = protractor.ExpectedConditions;

  this.editUserRole = async function(username) {
    await browser.get('/admin');
  
    var adminRolesTab = element(by.css('.protractor-test-admin-roles-tab'));
    await adminRolesTab.click();

    await expect(await adminRolesTab.getAttribute('class')).toMatch('active');

    var adminRolesTabContainer = element(by.cssContainingText('h1', 'Role Conainer'));
    await browser.wait(
      await until.visibilityOf(element),
      10000, 'Element not visible');

    var usernameInputFieldForRolesEditing = element.all(by.css(
      '.protractor-test-username-for-role-editor'));
    await usernameInputFieldForRolesEditing.first().sendKeys(username);

    var editUserRoleButton = await $('#protractor-test-button');
    await buttonText = editUserRoleButton.getText();

>     expect(buttonText).toBe('Button Text');

Reviewer: ```suggestion
    expect(buttonText).toBe('Button Text');
  }
```

Author: Done.

Reviewer: Resolved

## Table of contents

* [Overview](#overview)
* [Selectors](#selectors)
* [Events on selected elements](#events-on-selected-elements)
* [Points to note while migrating](#points-to-note-while-migrating)
* [Example migration](#example-migration)

## Overview

Currently, Oppia End-to-End tests are in a hybrid state where we have both WebdriverIO and Protractor test suites. The GSoC project **Migrate away from Protractor** aims to migrate all the existing End-to-End tests to WebdriverIO.

This guide is to help contributors in migrating end-to-end tests from Protractor to WebdriverIO during the hybrid state.

## Selectors

> A selector is used to query an element. The WebDriver Protocol provides several selector strategies to query an element. WebdriverIO simplifies them to keep selecting elements simple

Reviewer 2: You missed the addition of a period at the end from my [original comment](https://github.com/oppia/oppia-web-developer-docs/pull/7#discussion_r900411468):

```suggestion
A selector is used to query an element. The WebDriver Protocol provides several selector strategies to query an element. WebdriverIO simplifies them to keep selecting elements simple.
```

Author: Done.

| -----------------------------------------| -----------------------------------------|----------------------|
| Select element by class name             |`element(by.css('.some-css-class'))`           |`$('.some-css-class')`     |
| Select element containing certain string Text |`element(by.cssContainingText(tag, text))`|`$('tag=text')`       |
| Selecting Multiple Elements              |`element.all(by.css('.some-css-class'))`       |`$$('..some-css-class')`    |
| Select element by ID                     |`element(by.css('#some-css-id'))`              |`$('#some-css-id')`        |
| Select element inside a element with some class name |`element(by.css('some-css-class')).element(by.css('tag-within-css'))`  | `$('some-css-class').$('tag-within-css’)`      |
| Selecting element one by one in loop     |`element.get(i)`                          |`$$(selector)[i]`     |

There is no proper substitute available for the command `element.getWebElement()` in webdriverIO, but we can use `browser.findElement()` as a substitute for this.

For example, take look at the following code of protractor:

```js
 await browser.executeScript(

>     'arguments[0].click()', await clickableElement.getWebElement());

Reviewer 2: Indentation still not fixed ([original comment](https://github.com/oppia/oppia-web-developer-docs/pull/7#discussion_r900415005)). Also, `await` is indented unnecessarily by one space, which will annoy anyone trying to copy-paste this.

Ditto elsewhere

Author: Forgot to update it here, done now.

| Get an attribute from a DOM-element based on the attribute name.|`element.getAttribute(attributeName)`|`$(selector).getAttribute(attributeName)`|
| Wait until a given condition is fulfilled.|`browser.wait(condition, time, optional_msg)`|`browser.waitUntil(condition, { timeout, timeoutMsg, interval })`|
| Send a sequence of keystrokes to the active element|`element.sendKeys(keys)` |`element.setValue(keys)`     |
| Selected DOM-Element is clickable, visible, and exists.|`until.elementToBeClickable(element)` |`$(selector).isClickable()`|
|Get the value of a `<textarea>`, `<select>` or `<input>` found by given selector|`element.getText()` |`$(selector).getValue()`|
|Clear a `<textarea>` or `<input>` element’s value   |`element.clear()` |   `$(selector).clearValue()`  |

For more details please visit [webdriverio's official documentation](https://webdriver.io/docs/gettingstarted).

## Points to note while migrating

1. Currently, we are having two versions of the end-to-end test present in the codebase, so before moving forward with migration please refer to [Hybrid state doc](Hybrid-state.md) for a better understanding of this state.

> 2. The basic utility files i.e. forms.js, waitFor.js, action.js, general.js, user.js will already be migrated for the ease of contributors.

Reviewer 2: ```suggestion
3. The basic utility files i.e. forms.js, waitFor.js, action.js, general.js, user.js will already be migrated for the ease of contributors.
```

Author: Done.

## Overview

> Currently, Oppia E2E tests are in a hybrid state where we have both WebdriverIO and Protractor test suites. The GSoC project **Migrate away from Protractor** aims to migrate all the existing E2E tests to WebdriverIO.

Reviewer 2: Please spell out the acronym "E2E" for the first time you use it. Ditto elsewhere

Author: Done.

This guide is to help contributors in migrating end-to-end tests from Protractor to WebdriverIO during the hybrid state.

## Selectors

> A Selector is used to query an element. The WebDriver Protocol provides several selector strategies to query an element. WebdriverIO simplifies them to keep selecting elements simple

Reviewer 2: ```suggestion
A selector is used to query an element. The WebDriver Protocol provides several selector strategies to query an element. WebdriverIO simplifies them to keep selecting elements simple.
```

Please carefully check your PR for other grammar/capitalization/spelling errors like this one. I've noticed several already

Author: Done.

The `$` command is a short way to call the findElement command in order to fetch a single element on the page

| Goal                                     | Protractor                               |WebdriverIO           |
| -----------------------------------------| -----------------------------------------|----------------------|
| Select element by class name             |`element(by.css('.some-css-class'))`           |`$('..some-css-class')`     |
| Select element containing certain string Text |`element(by.cssContainingText(tag, text))`|`$('tag=text')`       |
| Selecting Multiple Elements              |`element.all(by.css('.some-css-class'))`       |`$$('..some-css-class')`    |
| Select element by ID                     |`element(by.css('#some-css-id'))`              |`$('#some-css-id')`        |

> | Select element inside a element with some class name |`element(by.css('some-css')).element(by.css('tag-within-css'))`  | `$('some-css').$('tag-within-css’)`      |

Reviewer 2: Why `some-css` here instead of `some-css-class` like you have elsewhere?

Author: Updated.

| Selecting element one by one in loop     |`element.get(i)`                          |`$$(selector)[i]`     |

There is no proper substitute available for the command `element.getWebElement()` in webdriverIO, but we can use `browser.findElement()` as a substitute for this.

For example, take look at the following code of protractor:

```js
 await browser.executeScript(

>       'arguments[0].click()', await clickableElement.getWebElement());

Reviewer 2: In JS we usually only indent by 2 spaces. ditto elsewhere

Author: Done.

| Get the text content from a DOM-element  |`element.getText())`                      |`$(selector).getText()`|
| Navigate to the given destination        |`browser.get(url)`                        |`browser.url(url)`     |
| Get the url of the currently opened website.|`browser.getCurrentUrl()`              |`browser.getUrl()`     |
| Get an attribute from a DOM-element based on the attribute name.|`element.getAttribute(attributeName)`|`$(selector).getAttribute(attributeName)`|
| Wait until a given condition is fulfilled.|`browser.wait(condition, time, optional_msg)`|`browser.waitUntil(condition, { timeout, timeoutMsg, interval })`|
| Send a sequence of keystrokes to the active element|`element.sendKeys(keys)` |`element.setValue(keys)`     |
| Selected DOM-Element is clickable, visible, and exists.|`until.elementToBeClickable(element)` |`$(selector).isClickable()`|
|Get the value of a `<textarea>`, `<select>` or `<input>` found by given selector|`element.getText()` |`$(selector).getValue()`|
|Clear a `<textarea>` or `<input>` element’s value   |`element.clear()` |   `$(selector).clearValue()`  |

> For more details please visit [ webdriverio ](https://webdriver.io/docs/gettingstarted) official documentation page.

Reviewer 2: ```suggestion
For more details please visit [webdriverio's official documentation](https://webdriver.io/docs/gettingstarted).
```

For accessibility, it's best to make the linked text describe the what the link goes to

Author: Done.

## Overview

> Currently, Oppia E2E tests are in a hybrid state where we have both WebdriverIO and Protractor test suites. The project **Migrate away from Protractor** aims to migrate all the existing E2E tests to WebdriverIO.

Reviewer 2: Maybe specify that "Migrate away from Protractor" is a GSoC project?

Author: Done.

> This guide is to help contributors write/modify E2E tests in webdriverIO during the hybrid state.

Reviewer: ```suggestion
This guide is to help contributors in migrating end to end tests from Protractor to WebdriverIO during the hybrid state.
```

Author: Done.

> This guide is to help contributors write/modify E2E tests in webdriverIO during the hybrid state.

Reviewer 2: ```suggestion
This guide is to help contributors write and modify E2E tests in webdriverIO during the hybrid state.
```

It's often unclear whether "/" means "and" or "or"

Author: Done.

> ## Selectors

Reviewer 2: Explain what a selector does

Author: Done.

The `$` command is a short way to call the findElement command in order to fetch a single element on the page

| Name                                     | Protractor                               |WebdriverIO           |
| -----------------------------------------| -----------------------------------------|----------------------|

> | By class name                            |`element(by.css('.className'))`           |`$('.className')`     |

Reviewer 2: It's unclear that `.className` should be replaced with the class name the user wants to select. Ditto throughout these tables--it's unclear what text the user is supposed to replace

Author: Updated.

| Element containing certain string Text   |`element(by.cssContainingText(tag, text))`|`$('tag=text')`       |
| Multiple Elements                        |`element.all(by.css('.className'))`       |`$$('.className')`    |
| By ID                                    |`element(by.css('#idName'))`              |`$('#idName')`        |

> | Chain Selectors                          |`element(by.css('some-css')).element(by.css('tag-within-css'))`  | `$('some-css').$('tag-within-css’)`      |

Reviewer 2: "Chain Selectors" isn't really a name of a selector type. Maybe the first column should instead describe what goal the selector accomplishes

Author: Done.

| Selecting element one by one in loop     |`element.get(i)`                          |`$$(selector)[i]`     |

> There is no proper substitute available for the command `element.getWebElement()` in webdriverIO but we can use `browser.findElement()` as a substitute for this.

Reviewer: May be you can show an eg of a slight code for` element.getWebElement()` (may be 2-3 lines that uses this code) in protractor where it's currently implemented and then show how it's going to be migrated in webdriverio. 
P.S I am not asking to link the code because that link might not work after the particular code is migrated.

Author: Done.

Reviewer: Resolved

## Actions

> | Actions                                  | Protractor                               |WebdriverIO            |

Reviewer 2: "Actions" might be confusing given that we have "action.js"

Author: Updated.

| -----------------------------------------| -----------------------------------------|-----------------------|
| Click on the element                     |`element.click()`                         |`$(selector).click()`  |
| Get the text content from a DOM-element  |`element.getText())`                      |`$(selector).getText()`|
| Navigate to the given destination        |`browser.get(url)`                        |`browser.url(url)`     |
| Get the url of the currently opened website.|`browser.getCurrentUrl()`              |`browser.getUrl()`     |
| Get an attribute from a DOM-element based on the attribute name.|`element.getAttribute(attributeName)`|`$(selector).getAttribute(attributeName)`|
| Wait until a given condition is fulfilled.|`browser.wait(condition, time, optional_msg)`|`browser.waitUntil(condition, { timeout, timeoutMsg, interval })`|

> | Send a sequence of keystrokes to the active element|`element.sendKeys(keys)` |`browser.keys(value`     |

Reviewer: I think [setvalue](https://webdriver.io/docs/api/element/setValue/) is a more vivid replacement. Can you please debug it?

Author: Yes, even tough element.keys are also working, and setValue is a proper replacement. Updated, thanks!

Reviewer: Resolved

| Selected DOM-Element is clickable, visible, and exists.|`until.elementToBeClickable(element)` |`$(selector).isClickable()`|

> |`Get the value of a <textarea>, <select> or text <input> found by given selector`|`element.getText()` |`$(selector).getValue()`|

Reviewer: Why have you put this in the coding style from github?
`Get the value of a <textarea>, <select> or text <input> found by given selector`. I think it would be better if you put the HTML tags.

Author: Done.

Reviewer: Resolved

> |`Clear a <textarea> or text <input> element’s value`   |`element.clear()` |   `$(selector).clearValue()`  |

Reviewer: Ditto

Author: Done.

Reviewer: Resolved

> |`Get the value of a <textarea>, <select> or text <input> found by given selector`|`element.getText()` |`$(selector).getValue()`|
> |`Clear a <textarea> or text <input> element’s value`   |`element.clear()` |   `$(selector).clearValue()`  |

Reviewer 2: ```suggestion
|Get the value of a `<textarea>`, `<select>` or text `<input>` found by given selector|`element.getText()` |`$(selector).getValue()`|
|Clear a `<textarea>` or text `<input>` element’s value   |`element.clear()` |   `$(selector).clearValue()`  |
```

Author: Done.

For more details please visit [ webdriverio ](https://webdriver.io/docs/gettingstarted) official documenation page.

> ## Points to Note While Migration

Reviewer 2: ```suggestion
## Points to note while migrating
```

We use sentence case in the wiki for headings (not title case)

Author: Done.

Reviewer 2: Not done (note `Migration` -> `migrating`)

Reviewer 2: Not done (note `Migration` -> `migrating`)

Author: Done now

> 1. `expect` calls must be prefixed with an `await`.

Reviewer: I am a bit confused why do we need this? Any specific reasons?

Author: Actually, this was meant for this [expect](https://webdriver.io/docs/api/expect-webdriverio/), I misunderstood this. Removed now.

Reviewer: Resolved.

> Example of **incorrect** code for this rule:

Reviewer 2: Please indent all the step 1 text to the same level as the first line (`expect` calls must be ...)

Author: Done.

Reviewer 2: Not done. You can see that this isn't right by looking at the rendered markdown https://github.com/oppia/oppia-web-developer-docs/blob/6d9f685778d96fbd72e8a0e9e6c8ac09c8f8202e/Guide-to-migrate-e2e-tests.md

Author: Now there is no need for indentation at this point as I have removed this.

```js
it('should visit the Get started page', async() => {
    await getStartedPage.get();
    expect(await $(
      '.webdribverio-test-get-started-page').isExisting()).toBeTrue();
});
```

Example of **correct** code for this rule:

```js
it('should visit the Get started page', async() => {
    await getStartedPage.get();
    await expect(await $(
      '.webdribverio-test-get-started-page').isExisting()).toBeTrue();
});
```

> 2. The basic utility files i.e forms.js, waitFor.js, action.js, general.js, user.js will already be migrated for the ease of contributors.

Reviewer: So, are you telling that it will be migrated at the start only?

Author: Yes, all these files will be migrated in the first few PR. (waitFor.js, action.js, general.js, user.js in the PR 1.2 and forms.js will be completed by 1.4)

Reviewer: Resolved

> 2. The basic utility files i.e forms.js, waitFor.js, action.js, general.js, user.js will already be migrated for the ease of contributors.

Reviewer 2: When will this be done? Will this page be updated once it's completed?

Author: waitFor.js, action.js, general.js, user.js in the PR 1.2 and forms.js will be completed by 1.4. Yes, I will update the wiki once it will be completed.

Reviewer 2: not done

Author: Done.

> 3. For more information on how to add new tests or modify the existing tests please refer to Write E2E tests section.

Reviewer: I think it would be more helpul if you can show an example of two versions of code in a seperate section in this file only. Need not to be a particular suite but something that covers most of the cases above.

Author: Done.

Reviewer: One would have been sufficient that could cover selectors and actions.

Author: Done.

Reviewer: I can now find some await statements before expect and element selectors. Can we not avoid it in order to maintain the original state that we have in protractor?

Author: Removed, yes they can work without await as well.

Reviewer: Resolved

> 3. For more information on how to add new tests or modify the existing tests please refer to Write E2E tests section.

Reviewer: Please link it to whatever you are referring to.

Author: Done.

Reviewer: action. zjs --> action. js

Also, an expected date would make it more clearer for the devs.

Author: Done.

Reviewer: Resolved

> 3. For more information on how to add new tests or modify the existing tests please refer to Write E2E tests section.

Reviewer 2: Not sure what the "Write E2E tests section" is. Also, can you provide a link?

Author: Done.
